### PR TITLE
ENH: Added new type `FeatureData[SequenceCharacteristics]` with semantic validator for property 'length'.

### LINKED ISSUE #324: ENH: Add new type `FeatureData[SequenceCharacteristics]`
- The addition is a new FeatureData type that will store characteristics of sequences. The first characteristic implemented will be the length of the sequence in basepairs. 
- The file stored will be a table with sequence identifiers as the index and then a column called "length" that stores the lengths of the sequences.
- The sequence lengths will be needed in normalisation actions in `q2-amr` and `q2-moshpit`.
- This type can be expanded in the future with additional characteristics that will correspond to additional properties and the columns can be validated with additional semantic validators.

This type was suggested in the [2024.05.02 Caporaso Bokulich group meeting](https://forum.qiime2.org/t/2024-05-02-group-meeting/30095)

### PR DESCRIPTION
**Pull request by VinzentRisch**
closes #324 

- Added new type `FeatureData[SequenceCharacteristics]`.
- Added new semantic validator for `FeatureData[SequenceCharacteristics % Properties("length")]`.

install this branch
```
git clone https://github.com/qiime2/q2-types.git
cd q2-types
git fetch origin pull/326/head:pr-326
git checkout pr-326
pip install -e .
```
test it out 
[sequence_characteristics.txt](https://github.com/qiime2/q2-types/files/15249255/sequence_characteristics.txt)
```
qiime tools import --input-path sequence_characteristics.txt --output-path  sequence_characteristics.qza --type "FeatureData[SequenceCharacteristics % Properties('length')]"
```

### CODE REVIEW

File: q2_types/feature_data/_transformer.py
Comment (misialq) [inline q2_types/feature_data/_transformer.py:1]: Could you also add a transformer to Q2 metadata?

File: q2_types/feature_data/_type.py [hunk 1]
```diff
@@ -52,13 +55,40 @@
                       variant_of=[FeatureData.field['type'],
                                   SampleData.field['type']])
 
+SequenceCharacteristics = SemanticType('SequenceCharacteristics',
+                                       variant_of=FeatureData.field['type'])
+
+
+@plugin.register_validator(FeatureData[SequenceCharacteristics %
+                                       Properties("length")])
+def validate_sequence_characteristics_length(data: pd.DataFrame, level):
```
Comment (misialq) [inline q2_types/feature_data/_type.py]: I don't want to be nit-picky but this function name is awfully long... Could we name it something like `validate_seq_char_len` maybe? 😅

**Review by misialq — CHANGES_REQUESTED** [review]
Hey @VinzentRisch, this mostly looks good to me - just one "small" request below. Thanks!

File: q2_types/feature_data/_type.py (hunk 1 above)
Comment (VinzentRisch) [reply q2_types/feature_data/_type.py]: Sure :)

File: q2_types/feature_data/_format.py [hunk 2]
```diff
@@ -470,6 +470,36 @@ def validate(self, *args):
...
+        except pd.errors.EmptyDataError:
+            raise ValidationError('File cannot be empty.')
+
+        if not data.columns.any():
+            raise ValidationError('File needs to have at least two columns.')
+
+
+SequenceCharacteristicsDirectoryFormat = model.SingleFileDirectoryFormat(
+    "SequenceCharacteristicsDirectoryFormat",
+    "sequence_characteristics.txt", SequenceCharacteristicsFormat
```
Comment (misialq) [inline q2_types/feature_data/_format.py]: It's a TAB-separated file so the extension should be tsv:
```suggestion
    "sequence_characteristics.tsv", SequenceCharacteristicsFormat
```

**Comment by misialq** [comment]
Hey @lizgehret, @ebolyen! This is the PR for the new type we were discussing in our last dev meeting - would one of you maybe have some time to check it out? 🙏

File: q2_types/feature_data/_transformer.py [hunk 3]
```diff
@@ -797,3 +798,22 @@ def _227(ff: BLAST6Format) -> qiime2.Metadata:
     # default int index but cast to a str and give it a name.
     data.index = pd.Index(data.index.astype(str), name='id')
     return qiime2.Metadata(data)
+
+
+@plugin.register_transformer
+def _228(ff: SequenceCharacteristicsFormat) -> pd.DataFrame:
+    return pd.read_csv(str(ff), sep="\t", index_col=0)
```
Comment (lizgehret) [inline q2_types/feature_data/_transformer.py]:
```suggestion
    return pd.read_csv(str(ff), sep='\t', index_col=0)
```

File: q2_types/feature_data/_transformer.py [hunk 4]
```diff
@@ -797,3 +798,22 @@ def _227(ff: BLAST6Format) -> qiime2.Metadata:
...
+@plugin.register_transformer
+def _229(data: pd.DataFrame) -> SequenceCharacteristicsFormat:
+    ff = SequenceCharacteristicsFormat()
+    data.to_csv(str(ff), sep='\t')
+    return ff
+
+
+@plugin.register_transformer
+def _230(ff: SequenceCharacteristicsFormat) -> qiime2.Metadata:
+    df = pd.read_csv(str(ff), sep="\t", index_col=0)
```
Comment (lizgehret) [inline q2_types/feature_data/_transformer.py]:
```suggestion
    df = pd.read_csv(str(ff), sep='\t', index_col=0)
```

File: q2_types/feature_data/_type.py [hunk 5]
```diff
@@ -120,3 +150,8 @@
     directory_format=BLAST6DirectoryFormat,
     description=("BLAST results associated with a set of feature "
                  "identifiers."))
+plugin.register_artifact_class(
+    FeatureData[SequenceCharacteristics],
+    directory_format=SequenceCharacteristicsDirectoryFormat,
+    description=("Characteristics of sequences (e.g., the length of a genes "
```
Comment (lizgehret) [inline q2_types/feature_data/_type.py]:
```suggestion
    description=("Characteristics of sequences (e.g., the length of a gene "
```

File: q2_types/feature_data/tests/test_format.py [hunk 6]
```diff
@@ -901,5 +903,37 @@ def test_blast6_format_invalid(self):
             BLAST6DirectoryFormat(temp_dir, mode='r').validate()
 
 
+class TestSequenceCharacteristicsFormat(TestPluginBase):
+    package = 'q2_types.feature_data.tests'
+
+    def test_sequence_characteristics_directory_format(self):
+        filepath = self.get_data_path('sequence_characteristics_length.txt')
```
Comment (lizgehret) [inline q2_types/feature_data/tests/test_format.py]:
```suggestion
        filepath = self.get_data_path('sequence_characteristics_length.tsv')
```
We probably want to rename these files with the `.tsv` extension, for consistency.

File: q2_types/feature_data/tests/test_format.py [hunk 7]
```diff
@@ -901,5 +903,37 @@ def test_blast6_format_invalid(self):
...
+        temp_dir = self.temp_dir.name
+        shutil.copy(filepath, os.path.join(temp_dir,
+                                           'sequence_characteristics.txt'))
```
Comment (lizgehret) [inline q2_types/feature_data/tests/test_format.py]:
```suggestion
                                           'sequence_characteristics.tsv'))
```

File: q2_types/feature_data/tests/test_format.py [hunk 8]
```diff
@@ -901,5 +903,37 @@ def test_blast6_format_invalid(self):
...
+        format = SequenceCharacteristicsDirectoryFormat(temp_dir, mode='r')
+        format.validate()
+
+    def test_sequence_characteristics_format(self):
+        filepath = self.get_data_path('sequence_characteristics_length.txt')
```
Comment (lizgehret) [inline q2_types/feature_data/tests/test_format.py]:
```suggestion
        filepath = self.get_data_path('sequence_characteristics_length.tsv')
```

File: q2_types/feature_data/tests/test_format.py [hunk 9]
```diff
@@ -901,5 +903,37 @@ def test_blast6_format_invalid(self):
...
+        format = SequenceCharacteristicsFormat(filepath, mode='r')
+        format.validate()
+
+    def test_sequence_characteristics_format_empty(self):
+        path = self.get_data_path('empty.txt')
```
Comment (lizgehret) [inline q2_types/feature_data/tests/test_format.py]:
```suggestion
        path = self.get_data_path('empty.tsv')
```

File: q2_types/feature_data/tests/test_format.py [hunk 10]
```diff
@@ -901,5 +903,37 @@ def test_blast6_format_invalid(self):
...
+        format = SequenceCharacteristicsFormat(path, mode='r')
+        with self.assertRaises(ValidationError) as context:
+            format.validate()
+        self.assertEqual(str(context.exception), 'File cannot be empty.')
+
+    def test_sequence_characteristics_format_only_index(self):
+        path = self.get_data_path('sequence_characteristics_only_index.txt')
```
Comment (lizgehret) [inline q2_types/feature_data/tests/test_format.py]:
```suggestion
        path = self.get_data_path('sequence_characteristics_only_index.tsv')
```

File: q2_types/feature_data/tests/test_transformer.py [hunk 11]
```diff
@@ -1507,5 +1507,43 @@ def test_blast6_to_metadata(self):
         assert_frame_equal(obs.to_dataframe(), exp)
 
 
+class TestSequenceCharacteristicsTransformer(TestPluginBase):
+    package = 'q2_types.feature_data.tests'
+
+    def setUp(self):
+        super().setUp()
+        self.exp_file = self.get_data_path(
+            "sequence_characteristics_length.txt")
```
Comment (lizgehret) [inline q2_types/feature_data/tests/test_transformer.py]:
```suggestion
            "sequence_characteristics_length.tsv")
```

File: q2_types/feature_data/tests/test_type.py [hunk 12]
```diff
@@ -118,6 +125,51 @@ def test_blast6_semantic_type_to_format_registration(self):
...
+                                            "empty (NaN) values.")
+
+    def test_validate_sequence_characteristics_length_negative_values(self):
+        data = self._setup_df()
+        data.loc[1, 'length'] = -1
+        self._assert_validation_error(data, "Column 'length' cannot contain "
+                                            "negative values.")
+
+    def _setup_df(self):
+        data_path = self.get_data_path("sequence_characteristics_length.txt")
```
Comment (lizgehret) [inline q2_types/feature_data/tests/test_type.py]:
```suggestion
        data_path = self.get_data_path("sequence_characteristics_length.tsv")
```

**Review by lizgehret — COMMENTED** [review]
overall this looks great - some veeery small/nitpicky comments (sorry!) but w/those suggestions it should be ready to merge!

File: q2_types/feature_data/_transformer.py (hunk 3 above)
Comment (lizgehret) [reply q2_types/feature_data/_transformer.py]: I know this is super nitpicky, but it really bothers me when single vs. double quotes aren't consistent (at least across individual files) 😝

File: q2_types/feature_data/tests/test_format.py (hunk 7 above)
Comment (lizgehret) [reply q2_types/feature_data/tests/test_format.py]: Ah, it also looks like there's a file name mismatch here - there isn't a `sequence_characteristics.tsv` file that's been added. Maybe this was supposed to be `sequence_characteristics_length.tsv`?

**Comment by VinzentRisch** [comment]
Hi @lizgehret 
Thanks for the review. 🙂
I now changed all quotes to be consistent with the other code.

**Review by misialq — APPROVED** [review]
All good on my side, thanks @VinzentRisch! Could you please still resolve the merge conflict?

**Comment by lizgehret** [comment]
changes lgtm, thanks @VinzentRisch and @misialq!

**Review by lizgehret — APPROVED** [review]

**Pull request merged**